packages/@react-aria/interactions/src/textSelection.ts

```diff
@@ -21,52 +21,71 @@ import {isIOS, runAfterTransition} from '@react-aria/utils';
 // There are three possible states due to the delay before removing user-select: none after
 // pointer up. The 'default' state always transitions to the 'disabled' state, which transitions
 // to 'restoring'. The 'restoring' state can either transition back to 'disabled' or 'default'.
+
+// For non-iOS devices, we apply user-select: none to the pressed element instead to avoid possible
+// performance issues that arise from applying and removing user-select: none to the entire page
+// (see https://github.com/adobe/react-spectrum/issues/1609).
 type State = 'default' | 'disabled' | 'restoring';
 
+// Note that state only matters here for iOS. Non-iOS gets user-select: none applied to the target element
+// rather than at the document level so we just need to apply/remove user-select: none for each pressed element individually
 let state: State = 'default';
 let savedUserSelect = '';
+let modifiedElementMap = new WeakMap<HTMLElement, string>();
 
-export function disableTextSelection() {
-  // Limit this behavior to iOS only. Android devices don't text select nearby element
-  // when long pressing on a different element.
-  if (!isIOS()) {
-    return;
-  }
+export function disableTextSelection(target?: HTMLElement) {
+  if (isIOS()) {
+    if (state === 'default') {
+      savedUserSelect = document.documentElement.style.webkitUserSelect;
+      document.documentElement.style.webkitUserSelect = 'none';
+    }
 
-  if (state === 'default') {
-    savedUserSelect = document.documentElement.style.webkitUserSelect;
-    document.documentElement.style.webkitUserSelect = 'none';
+    state = 'disabled';
+  } else if (target) {
+    // If not iOS, store the target's original user-select and change to user-select: none
+    // Ignore state since it doesn't apply for non iOS
+    modifiedElementMap.set(target, target.style.userSelect);
+    target.style.userSelect = 'none';
```

Comment:
I'm not sure if doing this directly on the element is a good idea. What if the element re-renders while it is pressed and `style` changes? React won't know about the modifications we've done here. Should we return a `style` object from `usePress` instead? Do we even need to add it only when pressed, or would it be fine to add it at all times? If only when pressed, we could probably use a regular state React update for that.

Reply:
Hm, good point, I can try refactoring this so that usePress returns a style object when `isPressed`. If I have it so that `user-select: none` is added to the pressable element at all times regardless of press state, it prevents users from selecting text via press and dragging (e.g. if they click+hold somewhere away from a link and drag past the link, it won't highlight the text)